### ISSUE TITLE
c++ concepts: "type constraint differs in template redeclaration"

### ISSUE THREAD

**Issue by nahratzah** [llvm/llvm-project#86500]
This is reproducible in both clang++17 and clang++18.
This used to work with clang++16.
(I verified using godbolt.)

I get this error:
```
<source>:10:18: error: type constraint differs in template redeclaration
   10 |         template<sender>
      |                  ^
<source>:15:30: note: in instantiation of template class 'some_template<int>::bar<int>' requested here
   15 | some_template<int>::bar<int> b;
      |                              ^
<source>:3:10: note: previous template declaration is here
    3 | template<sender>
      |          ^
1 error generated.
Compiler returned: 1
```

When compiling this code with clang++17:
```
template<typename> concept sender = true;

template<sender>
class foo;

template<typename X>  // <-- remove this and compiler is happy.
struct some_template {
    template<typename>  // <-- Or remove this and compiler is happy.
    class bar {
        template<sender>
        friend class ::foo;
    };
};

some_template<int>::bar<int> b;
```

The compiler is no longer unhappy, if I turn either `some_template` or `bar` into not-a-template (marked in the code).
Similarly if I use `typename` instead of a concept, the compiler is happy too.

**Comment by zyn0217** [comment]
https://godbolt.org/z/3Kdz39Wo5
I think this has been addressed by https://github.com/llvm/llvm-project/pull/84671, which has yet to be backported to 18.
cc @jcsxky in case you want.

**Comment by jcsxky** [comment]
> https://godbolt.org/z/3Kdz39Wo5 I think this has been addressed by #84671, which has yet to be backported to 18. cc @jcsxky in case you want.

Yeah, fixed by this pr with testing locally. @nahratzah You can cherry pick this commit and have a try.

**Comment by nahratzah** [comment]
Can confirm that fix addresses my issue, thank you. <3
